engine/sim/sc_expressions.cpp

```diff
@@ -1035,69 +1011,65 @@ bool is_binary( token_e expr_token_type )
 
 // parse_tokens =============================================================
 
-std::vector<expr_token_t> parse_tokens( action_t* action,
-                                        util::string_view expr_str )
+std::vector<expr_token_t> parse_tokens( action_t* action, util::string_view expr_str )
 {
   std::vector<expr_token_t> tokens;
-  // Make a copy to ensure the string is null-terminated.
-  std::string str( expr_str );
+  lexer_t lexer( expr_str );
 
-  expr_token_t token;
-  int current_index = 0;
-  token_e t         = TOK_UNKNOWN;
-
-  while ( ( token.type = next_token( action, str, current_index,
-                                     token.label, t ) ) != TOK_UNKNOWN )
+  token_e prev_token = TOK_UNKNOWN;
+  lexer_t::token_t token = lexer.next();
+  for (; token.type != TOK_EOF; token = lexer.next() )
   {
-    t = token.type;
-    tokens.push_back( token );
-  }
-
-  return tokens;
-}
+    if ( token.type == TOK_UNKNOWN )
+    {
+      if ( action )
+      {
+        action->sim->error( "{}-{}: Unexpected token ({}) in {}\n",
+                            action->player->name(), action->name(), token.str, expr_str );
+      }
+      else
+      {
+        fmt::print( "Unexpected token ({}) in {}\n", token.str, expr_str );
+      }
+      break;
+    }
 
-// print_tokens =============================================================
+    // lexer never produces binary SUB/ADD. convert unary MINUS/PLUS to them
+    // when they can "bind" to the previous token as a binary op
+    const bool prev_is_left_side = prev_token == TOK_STR || prev_token == TOK_NUM || prev_token == TOK_RPAR;
+    if ( token.type == TOK_MINUS && prev_is_left_side )
+      token.type = TOK_SUB;
+    if ( token.type == TOK_PLUS && prev_is_left_side )
+      token.type = TOK_ADD;
 
-void print_tokens( std::vector<expr_token_t>& tokens, sim_t* sim )
-{
-  std::string str;
-  if ( !tokens.empty() )
-    str += "tokens: ";
+    // optimization: fold minus followed by a number together into a single number token
```

Comment:
Couldn't this be done inside the lexer?
Maybe move the TOK_NUM matching (line 75 to 83) to a helper function, and then also call that when you have a TOK_MINUS match.

Reply:
I'd rather not tbh.
`-` can be both unary (`TOK_MINUS`) and binary (`TOK_SUB`). Right now the lexer does not care and always produces only the unary one. Moving this to the lexer will necessarily require the lexer to also handle binary `-` (as `)-1` should produce `[TOK_RPAR, TOK_SUB, TOK_NUM]`) which would require adding lookbehind.

I actually thought about handling `ceil` & `floor` outside of the lexer too (they can be easily matched outside with), don't remember why I decided to leave it here (the commit is from July when I was playing with an interpreter for expressions).

Reply:
Oh ok. So that's the main reason to restrict the lexer to unary is so that it works by just looking forward. And then it is easier to build binary SUB/ADD after the lexing pass and then optimize the leftover unary Minus+Num tokens.

Makes a lot of sense now that I fully understand it.

Reply:
Right, outside of "linguistic purity" reasons, this is mostly about making the lexer simpler.

It's mostly as a trade-off tbh. You still need to handle the "duality" of `-` (the fact that it's unary/binary depending on context) somewhere. It just so happens that handling it outside of the lexer is simpler - you have to handle the duality of `+` too and parsing is typically structured in such a way there is implicit "lookbehind".

If we used a more sophisticated parser this would be pretty obvious.

Reply:
Thanks again for all the explanations.